_posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd

```diff
@@ -0,0 +1,757 @@
+  recurrent_initializer = initializer_constant(value = 1)
+)
+lstm(input)
+
+# return_state = TRUE, return_sequences = TRUE
+# return shape = list of:
+#                - outputs, of shape: (batch_size, time steps, units)
+#                - "memory" states for the last time step, of shape: (batch_size, units)
+#                - "carry states" for the last time step, of shape: (batch_size, units)
+#
+# note how again, the "memory" state found in list item 2 matches the final-time step outputs reported in item 1
+lstm <- layer_lstm(
+  units = 1,
+  return_sequences = TRUE,
+  return_state = TRUE,
+  kernel_initializer = initializer_constant(value = 1),
+  recurrent_initializer = initializer_constant(value = 1)
+)
+lstm(input)
+```
+
+#### GRU
+
+```{r}
+# default args
+# return shape = (batch_size, units)
+gru <- layer_gru(
+  units = 1,
+  kernel_initializer = initializer_constant(value = 1),
+  recurrent_initializer = initializer_constant(value = 1)
+)
+gru(input)
+
+# return_sequences = TRUE
+# return shape = (batch_size, time steps, units)
+#
+# note how for each item in the batch, the value for time step 4 equals that obtained above
+gru <- layer_gru(
+  units = 1,
+  return_sequences = TRUE,
+  kernel_initializer = initializer_constant(value = 1),
+  recurrent_initializer = initializer_constant(value = 1)
+)
+gru(input)
+
+# return_state = TRUE
+# return shape = list of:
+#    - outputs, of shape: (batch_size, units)
+#    - "memory" states for the last time step, of shape: (batch_size, units)
+#
+# note how the list items are identical!
+gru <- layer_gru(
+  units = 1,
+  return_state = TRUE,
+  kernel_initializer = initializer_constant(value = 1),
+  recurrent_initializer = initializer_constant(value = 1)
+)
+gru(input)
+
+# return_state = TRUE, return_sequences = TRUE
+# return shape = list of:
+#    - outputs, of shape: (batch_size, time steps, units)
+#    - "memory states" for the last time step, of shape: (batch_size, units)
+#
+# note how again, the "memory state" found in list item 2 matches the final-time-step outputs reported in item 1
+gru <- layer_gru(
+  units = 1,
+  return_sequences = TRUE,
+  return_state = TRUE,
+  kernel_initializer = initializer_constant(value = 1),
+  recurrent_initializer = initializer_constant(value = 1)
+)
+gru(input)
+```
+
+### `torch`
+
+#### LSTM (non-stacked architecture)
+
+```{r}
+library(torch)
+
+# batch of 3, with 4 time steps each and a single feature
+# we will specify batch_first = TRUE when creating the LSTM
+input <- torch_randn(c(3, 4, 1))
+input
+
+# default args
+# return shape = (batch_size, units)
+#
+# note: there is an additional argument num_layers that we could use to specify a stacked LSTM - effectively composing two LSTM modules
+# default for num_layers is 1 though 
+lstm <- nn_lstm(
+  input_size = 1, # number of input features
+  hidden_size = 1, # number of hidden (and output!) features
+  batch_first = TRUE # for easy comparability with Keras
```

Comment:
This is not really really equivalent to keras, because keras has `bias=FALSE` by default. I know this doesn't change the point, so maybe just ignore that :)

Reply:
oh, but Keras has `use_bias = TRUE` it seems?